cpmpy/transformations/linearize.py

```diff
@@ -146,33 +157,108 @@ def linearize_constraint(lst_of_expr, supported={"sum","wsum"}, reified=False):
+                            raise ValueError("Attempting linerarization of unsafe division, safen expression first (cpmpy/transformations/safen.py)")
+
+                        r = intvar(*get_bounds(a % b)) # r is the remainder, reuse our bound calculations
+                        mult_res, side_cons = get_or_make_var(b * rhs)
+                        cpm_expr = eval_comparison(cpm_expr.name, a, mult_res + r)
+
+                        # need absolute values of variables later
+                        abs_of_a = intvar(*get_bounds(abs(a)))
+                        abs_of_b = intvar(*get_bounds(abs(b)))
+                        abs_of_rhs = intvar(*get_bounds(abs(rhs)))
+                        abs_of_r = intvar(*get_bounds(abs(r)))
+                        side_cons += [abs(a) == abs_of_a, abs(b) == abs_of_b, abs(rhs) == abs_of_rhs, abs(r) == abs_of_r]
+                        # |r| < |b|
+                        side_cons.append(abs_of_r < abs_of_b)
+
+                        # ensure we round towards zero
+                        mul_abs, extra_cons = get_or_make_var(abs_of_b * abs_of_rhs)
+                        side_cons += extra_cons + [mul_abs <= abs_of_a]
+                        newlist += linearize_constraint(side_cons, supported=supported, reified=reified)
 
                 else:
                     raise TransformationNotImplementedError(f"lhs of constraint {cpm_expr} cannot be linearized, should"
                                                             f" be any of {supported | {'sub'} } but is {lhs}. "
                                                             f"Please report on github")
 
-            elif isinstance(lhs, GlobalConstraint) and lhs.name not in supported:
+            elif isinstance(lhs, GlobalFunction) and lhs.name == "abs" and abs not in supported:
+                if cpm_expr.name != "==": # TODO: remove this restriction, requires comparison flipping
+                    newvar = intvar(*get_bounds(lhs))
+                    newlist += linearize_constraint([lhs == newvar])
+                    cpm_expr = eval_comparison(cpm_expr.name, newvar, rhs)
+                else:
+                    x = lhs.args[0]
+                    lb, ub = get_bounds(x)
+                    if lb >= 0:  # always positive
+                        newlist.append(x == rhs)
+                    elif ub <= 0:  # always negative
+                        newlist.append(x + rhs == 0)
+                    else:
+                        lhs_is_pos = cp.boolvar()
+                        newcons = [lhs_is_pos.implies(x >= 0), (~lhs_is_pos).implies(x <= -1),
+                                   lhs_is_pos.implies(x == rhs), (~lhs_is_pos).implies(x + rhs == 0)]
+                        newlist += linearize_constraint(newcons, supported=supported, reified=reified)
+                    continue # all should be linear now
+
+
+            elif isinstance(lhs, GlobalFunction) and lhs.name not in supported:
```

Comment:
should this be globalconstraint? globalfunction does not inherit from globalconstraint anymore so maybe should be both?

Reply:
Ah i see it's still at the bottom of the file, it looks deleted here